src/polyswarm_api/types/artifact.py

```diff
@@ -0,0 +1,277 @@
+from ordered_set import OrderedSet
+
+from ..exceptions import NotFoundException, InvalidArgument, ArtifactDeleted
+from ..log import logger
+from .base import BasePSJSONType, ArtifactType
+from .hash import all_hashes, Hash, Hashable
+from io import BytesIO
+import os
+from . import schemas
+from .scan import Bounty
+
+
+def requires_analysis(func):
+    def wrapper(a, *args, **kwargs):
+        if not a.analyzed:
+            a.analyze_artifact()
+        return func(a, *args, **kwargs)
+    return wrapper
+
+
+def not_deleted(func):
+    def wrapper(a, *args, **kwargs):
+        if a.deleted:
+            raise ArtifactDeleted("Tried to use deleted LocalArtifact")
+        return func(a, *args, **kwargs)
+    return wrapper
+
+
+class ArtifactInstance(BasePSJSONType):
+    SCHEMA = schemas.artifact_instance_schema
+
+    def __init__(self, artifact, json, polyswarm=None):
+        super(ArtifactInstance, self).__init__(json, polyswarm)
+        self.artifact_id = json['artifact_id']
+        self.bounty_id = json['bounty_id']
+        # rename this to make it clearer, and avoid the word 'Result' which has a specific meaning in this project
+        # using Scan here instead of Bounty
+        self.bounty = Bounty(self, json['bounty_result']) if json['bounty_result'] else None
+        self.community = json['community']
+        self.consumer_guid = json.get('consumer_guid', None)
+        self.country = json['country']
+        self.id = json['id']
+        self.name = json['name']
+        self.submitted = json['submitted']
+        self.artifact = artifact
+
+
+class Artifact(Hashable, BasePSJSONType):
+    SCHEMA = schemas.artifact_schema
+
+    def __init__(self, json, polyswarm=None):
+        """
+        A representation of artifact data retrieved from the polyswarm API
+
+
+        :param path: Path to the artifact
+        :param content: Content of the artifact
+        :param artifact_name: Name of the artifact (filename, or otherwise)
+        :param artifact_type: ArtifactType of the artifact
+        :param polyswarm: Current PolyswarmAPI instance
+        :param json: JSON used to
+        :param analyze:
+        """
+        super(Artifact, self).__init__(json, polyswarm)
+
+        self.mimetype = json['mimetype']
```

Comment:
I like that we are trying to use objects rather than a simple json here. However, this creates 2 "thuths" about what comes back from the `API`. Suppose someday we change the `mimetype` field to `mtype`. This would require lots of changes in other places and would be a maintenance burden. Also, if we add or delete things, these don't automatically reflect on this object. I would therefore, change how this is done and keep a single truth, the one that the `API` returns (not tested):
```python
for key, value in json.items():
    setattr(self, key, value)
```
Doing type parsing like this is harder, but we get instantaneous changes in the object once the API changes. We could provide a list of parsers for some keys in order to do some default parsing in case it is present, but which would not break in case it is not found:
```python
self.hashes = ['sha256', 'sha1', 'md5']
for key, value in json.items():
    if key in self.hashes:
        value = Hash(value, key, polyswarm)
    setattr(self, key, value)
```
This is just an example of how to parse, we could create a hash of parsers and have them applied in a more generic way for sure, but these would be "nice to have parsings" rather than "enforced field access with static names".

Reply:
My reason for not doing this here is not that I don't like less work, but that this doesn't really absolve us of work if we want these objects to be usable by mere mortals. Instead of having the attributes defined in code, now we would have to go and make sure that we properly document in the class itself what attributes people should expect in this class. If we have to go and document all of these fields anyway, why not make them first class citizens?

Maybe there is some in-between ground here to lessen the overhead, but I'd like to avoid dynamically setting all of these for now I think.

Reply:
Hum... I think the definition of what exists in the api should exist in a single place, and that would be the api documentation. Users should look up there which fields they can access. Given that, once we materialize them, we are creating two definitions, one from the api and one from the class. 

Once we have two definitions, we are bound to have client/api synchronization problems. If a user changes the version of the client, it might immediately become incompatible with the api. We will probably fall into the temptation of making quick fixes like `self.some_old_name = api['some_new_name']` to keep things backwards compatible and other issues that we should not really support in the client, but rather supporting multiple api versions. 

As a solution, I would add a doc string that points to the permalink that describes the api for this resource. That way, users could check the current documentation for the fields that are available (if we don't have a permalink then that is something we should work on). I believe users would understand it easily and it is maintainable over time. To be fair, there is one reason that makes it interesting to have it as a first-class attribute, that is code completion. However, I really think it will create more issues than help people. A well written documentation would be much better than a client that breaks when upgraded in my opinion.

Since this is the kind of change that is easy to make later, I am fine with keeping things as they are and, if we realize we need it , we can change in the future. I believe that having the dynamic binding now would push us towards having a better documentation sooner rather than later though. I would add some comments to remind people that we should try to create attributes with the same name that it is given in the api key though. Also, I would try not to transform what is in each of these keys in order to make things easy to switch if we need it one day.

Reply:
The other potential issue here is breaking client code through API changes. Until we have a better story for versioning, I really don't want to risk us removing some field in our API, only to suddenly brick a client's infrastructure. Anyway, we can discuss it more, but yeah, I think this needs to be in the future column.